### ISSUE TITLE
Feature request: @import-once

### ISSUE THREAD

**Issue by seanofw** [sass/sass#156]
I've used Sass on a number of projects now with great success, and finally convinced my company to use it on our flagship product --- a _massive_ web application --- to make our 120,000 lines of CSS more manageable.

In doing so, I've found that there's a feature that would make for a really nice addition:  The ability to mark some files with an "@import-once" directive, which causes that file to be imported only once no matter how many different other files import it.

Consider this scenario:
- __SharedBaseStuff.scss_ has a bunch of base definitions.
- __FooWidget.scss_ imports __SharedBaseStuff.scss_
- __BarWidget.scss_ imports __SharedBaseStuff.scss_
- _YourPage.scss_ imports __FooWidget.scss_
- _MyPage.scss_ imports __BarWidget.scss_
- _TheirPage.scss_ imports both __FooWidget.scss_ and __BarWidget.scss_

Without doing anything else, TheirPage.css will end up with two copies of any CSS found in _SharedBaseStuff.scss, which is undesirable.  It'd be far better if inside _SharedBaseStuff.scss, you could just declare @import-once; at the top, which would ensure you only get the first copy in the output no matter how many times it's imported.  (By definition, @import-once could not be appear inside an imported file that is nested within another selector or mixin; it could only appear at the root.)

(Yes, this is a real scenario on our huge site.  In fact, our real-world scenarios are far uglier, with some .scss files able to be imported dozens of times or more by a single page's stylesheet.)

Currently, a little function like this can be used to implement a workaround:

```
$imported-once-files: ();

@function import-once($name) {
    @if index($imported-once-files, $name) {
        @return false;
    }
    $imported-once-files: append($imported-once-files, $name);
    @return true;
}
```

And then used like this:

```
@if import-once("_SharedBaseStuff.scss") {

    ...declare stuff that will only be imported once...

}
```

That works, but for large filesets with lots of imports, it's inefficient.  Having a simple global @import-once; directive would make this much simpler.

**Comment by nex3** [comment]
This is something we'll consider for the upcoming re-envisioning of the `@import` functionality.

**Comment by chriseppstein** [comment]
Duplicate of https://github.com/nex3/sass/issues/139

**Comment by chriseppstein** [comment]
BTW, that @if stuff is pretty clever :) It's pretty much how C solves this problem.

**Comment by seanofw** [comment]
Yah, I've done a lot of C coding, give or take a million lines. :-)

For the record, I think it's more meaningful to have the once-ness determined by the importee rather than the importer --- by the file that's being imported rather than by the file that's importing it.  @import-once is a safe construct; no matter who imports that file, they can't screw it up, because it's impossible to import that file more than once; the caller can't make a mistake.  But the suggestion in the other thread (which I missed when I searched for this topic), where you might add @require --- that suggestion allows the _caller_ to decide whether the file gets imported once or many times, which is dangerous:  If any one caller forgets and uses @import instead of @require, the imported code may end up imported more than once because of it.  I've been bit by that in other languages, and I'd rather not see the same problem repeated in Sass.

**Comment by davidrivers** [comment]
+1 for native support. I just had a specificity-override scenario where I had to implement the import-once @function above to work around it :(

**Comment by JawsomeJason** [comment]
+1 for native support.  I'd like to use the above workaround for magic sprites, but the "@include" portion cannot live inside an @if.

**Comment by chriseppstein** [comment]
Now available as a plugin: https://github.com/chriseppstein/compass/blob/master/import-once/README.md

**Comment by jlong** [comment]
@chriseppstein Did this get merged into 3.3 or is it still just a plugin?

**Comment by cseeger** [comment]
@jlong, it looks like this will be part of the 4.0 release.

**Comment by emilypriddy** [comment]
That will be a great addition!

**Comment by bobber205** [comment]
Years later this basic functionality still isn't added? :(
